dora/stdlib/Zero.dora

```diff
@@ -0,0 +1,59 @@
+// Implementations of the `Zero` trait indicate that zero-initialized memory of the appropriate size can be allocated to
+// represent a valid value of that type.¹
+//
+//   let numbers: Array[Int] = Array.ofZero[Int](4)
+//   assert(numbers == Array(0, 0, 0, 0)
+//
+//   let strings: Array[Option[String]] = Array.ofZero[Option[String]](2)
+//   assert(strings == Array(None, None)
+//
+// `Zero` is a special trait whose implementations are automatically derived for qualifying types.
+// It cannot be implemented manually.
+// A type qualifies if it is either `Option`, or obeys all of the following rules ...
+// - The type is a struct.
+// - The type does not contain any reference-types directly or transitively.
+//
+// ¹ This means that implementations of `Zero` may make full use of the operating system's optimization of `calloc`
+//   (lazy allocation, lazy zeroing, copy on write, etc.), unlike implementations of the `Default` trait –
+//   which invoke an user-defined function to initialize each individual element (malloc & loop with function calls).
+
+
+// @restricted / @internal
+trait Zero {
+  @static fun zero() -> Self; // should be `let` instead of `fun`
+}
+
+impl Zero for Bool {
```

Comment:
I think we should implement the trait for all possible types automatically in the runtime in the future (no impl for Zero in the source). For this PR I think we can keep it this way, since this already allows us to use `arrayZero` without additional work.

Reply:
> I think we should implement the trait for all possible types automatically in the runtime in the future (no impl for Zero in the source).

Yes, that was my plan!

> For this PR I think we can keep it this way, since this already allows us to use arrayZero without additional work.

Agreed.